layabout.py

```diff
@@ -1,3 +1,287 @@
+
+
+class MissingSlackToken(LayaboutError):
+    """ Raised if a Slack API token could not be found. """
+
+
+class FailedConnection(LayaboutError, ConnectionError):
+    """
+    Raised if the Slack client could not connect to the Slack API.
+
+    Inherits from both :obj:`LayaboutError` and the built-in
+    :obj:`ConnectionError` for convenience in exception handling.
+    """
+
+
+class Layabout:
+    """
+    An event handler on top of the Slack RTM API.
+
+    Args:
+        name: A unique name for this :obj:`Layabout` instance.
+        env_var: The environment variable to try to load a Slack API token
+            from.
+
+    Attributes:
+        name (str): A unique name for this :obj:`Layabout` instance.
+        slack (SlackClient): A :obj:`slackclient.SlackClient` instance.
+            Initially :obj:`None`, the attribute is set after calling the
+            :obj:`Layabout.connect` or :obj:`Layabout.run`
+            methods.
+
+    Example:
+
+        .. code-block:: python
+
+           from layabout import Layabout
+
+           layabout = Layabout('app')
+
+
+           @layabout.handle('message')
+           def echo(slack, event):
+               \"\"\" Echo all messages seen by the app. \"\"\"
+               channel = event['channel']
+               message = event['text']
+               subtype = event.get('subtype')
+
+               # Avoid an infinite loop of echoing our own messages.
+               if subtype != 'bot_message':
+                   slack.rtm_send_message(channel, message)
+
+           layabout.run()
+    """
+    def __init__(self, name: str, env_var: str = 'SLACK_API_TOKEN') -> None:
+        # TODO: Consider keyword only arguments.
+        self.name = name
+        self._env_var = env_var
+        self._token: str = None
+        self.slack: SlackClient = None
+        self._handlers: _Handlers = defaultdict(list)
+
+    def handle(self, type: str, kwargs: dict = None) -> Callable:
+        """
+        Register an event handler with the :obj:`Layabout` instance.
+
+        Args:
+            type: The name of a Slack RTM API event to be handled. As a
+                special case, although it is not a proper RTM event, ``*`` may
+                be provided to handle all events. For more information about
+                available events see the
+                `Slack RTM API <https://api.slack.com/rtm>`_.
+            kwargs: Optional arbitrary keyword arguments passed to the event
+                handler when the event is triggered.
+
+        Returns:
+            A decorator that validates and registers a Layabout event handler.
+
+        Raises:
+            TypeError: If the decorated :obj:`Callable`'s signature does not
+                accept at least 2 parameters.
+        """
+        def decorator(fn: Callable) -> Callable:
+            # Validate that the wrapped callable is a suitable event handler.
+            sig = signature(fn)
```

Comment:
I totally don't know what this does. My guess is that it generates metadata about the function?

Reply:
Yes, it returns an [`inspect.Signature`](https://docs.python.org/3/library/inspect.html#inspect.Signature) which includes, among other things, the number of parameters and their names.